### PR TITLE
Update precommit

### PR DESCRIPTION
**Pull request by EarlMilktea**
This PR updates pre-commit CI for better style checking.

### Description of changes

- Upgrade hooks in `.pre-commit-config.yaml`
- Remove deprecated `mirrors-autopep8`
- Unpin `types-setuptools` because `setuptools` itself is not pinned
- Fix linter warnings
  - ~Ignoring rule E721 for now~
- Change/upgrade stylecheck deps. (deprecated)
  - Drop `pycodestyle` as it's shipped with `autopep8`

### CODE REVIEW

File: cupyx/scipy/sparse/linalg/_iterative.py [hunk 1]
```diff
@@ -339,7 +339,6 @@ def _make_fast_matvec(A):
     from cupy_backends.cuda.libs import cusparse as _cusparse
     from cupyx import cusparse
 
-    matvec = None
```
Comment (asi1024) [inline cupyx/scipy/sparse/linalg/_iterative.py:342]: `None` should be returned if `_csr.isspmatrix_csr(A) and cusparse.check_availability('spmv')` is False.

Comment (EarlMilktea) [reply cupyx/scipy/sparse/linalg/_iterative.py:342]: Fixed in f7219c58432367e48ccb65bdfc8d55ed35235c50 .

File: cupyx/scipy/sparse/linalg/_iterative.py [hunk 2]
```diff
@@ -1,10 +1,10 @@
-import numpy
-
 import cupy
+import numpy
```
Comment (asi1024) [inline cupyx/scipy/sparse/linalg/_iterative.py:2]: Could you roll back this modification in accordance with rule H306?

```
Import statements must be organized into three parts: standard libraries, third-party libraries, and internal imports. [H306]
```
https://docs.cupy.dev/en/latest/contribution.html#coding-guidelines

Comment (EarlMilktea) [reply cupyx/scipy/sparse/linalg/_iterative.py]: Fixed in d2f93976a56c33f8440c637a3cb568a34787bf8f .
(This modification was accidentally inserted by vscode)

**Review by asi1024 — APPROVED** [review]

**Comment by asi1024** [comment]
LGTM!

**Pull request merged**